java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql

```diff
@@ -0,0 +1,93 @@
+import semmle.code.java.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/** The Java class `java.security.MessageDigest` */
+class MessageDigest extends RefType {
+  MessageDigest() { this.hasQualifiedName("java.security", "MessageDigest") }
+}
+
+/** The method `digest()` declared in `java.security.MessageDigest`. */
+class MDDigestMethod extends Method {
+  MDDigestMethod() {
+    getDeclaringType() instanceof MessageDigest and
+    hasName("digest")
+  }
+}
+
+/** The method `update()` declared in `java.security.MessageDigest`. */
+class MDUpdateMethod extends Method {
+  MDUpdateMethod() {
+    getDeclaringType() instanceof MessageDigest and
+    hasName("update")
+  }
+}
+
+/**
+ * Gets a regular expression for matching common names of variables that indicate the value being held is a password.
+ */
+string getPasswordRegex() { result = "(?i).*pass(wd|word|code|phrase).*" }
+
+/** Finds variables that hold password information judging by their names. */
+class PasswordVarExpr extends Expr {
+  PasswordVarExpr() {
+    exists(Variable v | this = v.getAnAccess() | v.getName().regexpMatch(getPasswordRegex()))
+  }
+}
+
+/** Taint configuration tracking flow from an expression whose name suggests it holds password data to a method call that generates a hash without a salt. */
+class HashWithoutSaltConfiguration extends TaintTracking::Configuration {
+  HashWithoutSaltConfiguration() { this = "HashWithoutSaltConfiguration" }
+
+  override predicate isSource(DataFlow::Node source) { source.asExpr() instanceof PasswordVarExpr }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(
+      MethodAccess mda, MethodAccess mua // invoke `md.digest()` with only one call of `md.update(password)`, that is, without the call of `md.update(digest)`
+    |
+      sink.asExpr() = mda.getQualifier() and
+      mda.getMethod() instanceof MDDigestMethod and
+      mda.getNumArgument() = 0 and // md.digest()
+      mua.getMethod() instanceof MDUpdateMethod and // md.update(password)
+      mua.getQualifier() = mda.getQualifier().(VarAccess).getVariable().getAnAccess() and
+      not exists(MethodAccess mua2 |
+        mua2.getMethod() instanceof MDUpdateMethod and // md.update(salt)
+        mua2.getQualifier() = mua.getQualifier().(VarAccess).getVariable().getAnAccess() and
+        mua2 != mua
+      )
+    )
+    or
+    // invoke `md.digest(password)` without another call of `md.update(salt)`
+    exists(MethodAccess mda |
+      sink.asExpr() = mda and
+      mda.getMethod() instanceof MDDigestMethod and // md.digest(password)
+      mda.getNumArgument() = 1 and
+      not exists(MethodAccess mua |
+        mua.getMethod() instanceof MDUpdateMethod and // md.update(salt)
+        mua.getQualifier() = mda.getQualifier().(VarAccess).getVariable().getAnAccess()
+      )
+    )
+  }
+
+  /** Holds for additional steps such as `passwordStr.getBytes()` */
+  override predicate isAdditionalTaintStep(DataFlow::Node pred, DataFlow::Node succ) {
```

Comment:
This is very general, propagating all method arguments to all results and qualifiers -- suggest restricting this to particular cases you're interested in.

Reply:
Added the type constraint

Reply:
So it's conducting taint from the arg of a digest or update to the hash accumulator? Why do we want that in the context of tracking taint from a password to a digest call? Could you add a test that only works with the additional-taint-step implemented, or indicate one that requires it now?

Reply:
Without the additional taint step, somehow no correct result will be returned. So there must be something wrong with the logic. I will further investigate.

Reply:
Sorry I confused myself. I removed the specific method names and just specified the object type. 

This additional taint step is for the check of:
```
// invoke `md.digest()` with only one call of `md.update(password)`, that is, without the call of `md.update(digest)`
```

Without it, the following invocation won't be detected since it's a separate method access:
```
	// BAD - Hash without a salt.
	public void getSHA256Hash2(String password) throws NoSuchAlgorithmException {
		MessageDigest md = MessageDigest.getInstance("SHA-256");
		md.update(password.getBytes());
		byte[] messageDigest = md.digest();
	}
```

And the part <code>pred.asExpr() = ma.getAnArgument()</code> also helps since the sink below is the method access itself instead of its argument:
```
    // invoke `md.digest(password)` without another call of `md.update(salt)`
    exists(MethodAccess mda |
      sink.asExpr() = mda and
```

Please let me know if we can further improve it. Thanks.

Reply:
Isn't that only the case because you define `sink.asExpr() = mda.getQualifier()`? What if the sink was the argument to the `update` call, then the rest of the definition of an interesting update call (looking for sibling digest() and other update calls) would do the rest of the work?

Reply:
Thanks a lot. That helps to remove the additional taint step.